doc/wrap_native_functions.md

```diff
@@ -0,0 +1,209 @@
+The `variants` field tells that `mm` has two forms: a C++ global function and
+the method of C++ class `at::Tensor`, where `at` stands for ATen, which is
+another sub-package that defines the fundamental data type `Tensor`.
+
+This tutorial covers the wrapping of the form of global function.  Readers are
+welcome to take the wrapping of the method form as an exercise.
+
+The PyTorch build system generates declarations of global function forms in
+`ATen/Functions.h`.  Download and unzip the official release of libtorch zip
+[archives](https://pytorch.org/get-started/locally/), you will find the file as
+`libtorch/include/ATen/Functions.h`.  The declaration of the native function
+`mm` is as the following.
+
+```cpp
+namespace at {
+CAFFE2_API Tensor mm(const Tensor & self, const Tensor & mat2);
+}  // namespace at
+```
+
+The method `at::Tensor::mm` is in `libtorch/include/ATen/core/TensorBody.h`.
+
+```cpp
+namespace at {
+class CAFFE2_API Tensor {
+  Tensor mm(const Tensor & mat2) const;
+ protected:
+  c10::intrusive_ptr<TensorImpl, UndefinedTensorImpl> impl_;
+};
+}  // namespace at
+```
+
+The only data member of class `at::Tensor` is `impl_`, a smart pointer pointing
+to an object that implements the details.
+
+## C Wrappers and C Tensor
+
+The native functions are in C++, and Go can only call C but not C++ functions,
+so we need to write a C wrapper for each native function.
+
+There are additional reasons for C wrappers:
+
+1. If a native function returns a tensor, its C wrapper moves the tensor into an
+   object in the heap, so it will not be free up by C++ smart pointers so that
+   the Go code can use it.
+1. Encapsulate the C++ class `at::Tensor` by a C type that can be used by Go
+   code.
+1. The native functions might throw C++ exceptions.  The C wrappers convert
+   exceptions into C strings, which, in turn, converted by the Go wrapper into
+   Go panics.
+
+With Cgo, Go programs can refer to C symbols with their names prefixed by `C.`.
+For example, the following Go function `MyExit` calls the C standard function
+`exit` declared in `stdlib.h` as `C.exit`.
+
+```go
+// #include <stdlib.h>
+import "C"
+func MyExit(x int) {
+    C.exit(x)
+}
+```
+
+We put all C wrappers of native functions in the subdirectory `cgotorch`.  In
+`cgotorch/cgotorch.h`, we can see the wrapper of `at::Tenosr` and `at::mm`.
+
+```c
+extern "C" {
+typedef at::Tensor* Tensor;
+const char *MM(Tensor a, Tensor b, Tensor *result);
+}
+```
+
+C does not have classes, but C has pointers, so we define C type `Tenosr` as a
+pointer to `at::Tensor`.  Go programs can use C pointers as of type
+`unsafe.Pointer`.
+
+The C wrapper `MM` returns a string serialization of the possible C++ exception
+thrown by libtorch if there is any, or `nullptr`.
+
+The implementation of the C wrapper `MM` is in C++ and in `cgotorch/torch.cc`.
+
+```cpp
+const char *MM(Tensor a, Tensor b, Tensor *result) {
+  try {
+    at::Tensor c = at::mm(*a, *b);
+    *result = new at::Tensor(c);
```

Comment:
No. It doesn't, because here the new operator calls the [move constructor](https://en.cppreference.com/w/cpp/language/move_constructor) instead of the copy constructor. Let me explain this in the doc.